pybio/spec/schema.py

```diff
@@ -0,0 +1,250 @@
+from dataclasses import asdict
+from typing import Any, Dict, Union
+
+from marshmallow import Schema as _Schema, pprint, ValidationError, post_load, validates_schema, validates
+
+from pybio.spec import pybio_types, fields
+from pybio.spec.pybio_types import MagicTensorsValue, MagicShapeValue
+
+
+class Schema(_Schema):
+    # class Meta:
+    # unknown = EXCLUDE
+    @post_load
+    def make_object(self, data, **kwargs):
+        if not data:
+            return None
+
+        this_type = getattr(pybio_types, self.__class__.__name__)
+        try:
+            return this_type(**data)
+        except TypeError as e:
+            e.args += (f"when initializing {this_type} from {self}",)
+            raise e
+
+
+class CiteEntry(Schema):
+    text = fields.Str(required=True)
+    doi = fields.Str(missing=None)
+    url = fields.Str(missing=None)
+
+    @validates_schema
+    def doi_or_url(self, data, **kwargs):
+        if data["doi"] is None and data["url"] is None:
+            raise ValidationError("doi or url needs to be specified in a citation")
+
+
+class MinimalYAML(Schema):
```

Comment:
Will it be confusing at some point that there is a schema and dataclass named the same? Maybe we're never using them together?

Reply:
at the moment they are all mirrored (there is also code for autogenerating schemas from dataclasses: https://pypi.org/project/marshmallow-dataclass/, but not all fields would map one to one)